File: 0157-read-n-characters-given-read4/0157-read-n-characters-given-read4.java
/**
 * The read4 API is defined in the parent class Reader4.
 *     int read4(char[] buf4);
 */

public class Solution extends Reader4 {
    /**
     * @param buf Destination buffer
     * @param n   Number of characters to read
     * @return    The number of actual characters read
     */
    public int read(char[] buf, int n) {
        for (int i=0;i*4<n;i++) {
            char[] buf4 = new char[4];
            int charRead = read4(buf4);
            for(int j = i*4,k=0;k<charRead && i*4+k<n;j++,k++) {
                buf[j] = buf4[k];
            }
            if(charRead < 4) {
                return Math.min(i*4+charRead,n);
            }
        }
        return n;
    }
}

File: 0157-read-n-characters-given-read4/0157-read-n-characters-given-read4.cpp
/**
 * The read4 API is defined in the parent class Reader4.
 *     int read4(char *buf4);
 */

class Solution {
public:
    /**
     * @param buf Destination buffer
     * @param n   Number of characters to read
     * @return    The number of actual characters read
     */
    int read(char *buf, int n) {
        int ans = 0;
        char buf4[4];
        
        for(int i=0;i<=(n-1)/4;i++){
            int curr = read4(buf4);
            if(curr<=0)
                return ans;
            for(int j=0;j<curr && i*4+j<n;j++) {
                buf[i*4+j] = buf4[j];
            }
            ans+=curr;
        }
        return ans;
    }
};